File: agent-skills/archive-fact/scripts/archive-fact.mjs
#!/usr/bin/env node
// archive-fact.mjs — safely MOVE a decayed fact from continuity.md to the archive.
//
// Node port of archive-fact.py, kept at output parity (shared test fixtures are the
// cross-runtime contract). Executes the mechanical half of REVIEW.md step 4: the agent
// decides *what* to archive (meaning); this performs the *move* safely (mechanics).
//
// Why: `open(f,"w").write(open(f).read()+…)` truncates before the read and wiped this
// repo's archive once already. This reads the whole file into memory FIRST and writes
// once, so truncation is structurally impossible.
//
// Usage: node archive-fact.mjs --id ID [--id ID2 ...] [--reason REASON]
//                              [--quarter YYYY-Qn] [--note TEXT] [--root PATH] [--dry-run]
// Exit:  0 = moved (or dry-run ok), 1 = refused, 2 = could not locate memory/.
//        Run `memory-lint` afterward to confirm.

import { readFileSync, writeFileSync, appendFileSync, existsSync, statSync, readdirSync } from "node:fs";
import { resolve, dirname, join, basename } from "node:path";
import { fileURLToPath } from "node:url";

export function find_root(start) {
  const here = dirname(fileURLToPath(import.meta.url));
  for (const cand of [start, process.cwd(), here]) {
    if (!cand) continue;
    let d = resolve(cand);
    while (true) {
      const f = join(d, "memory", "continuity.md");
      if (existsSync(f) && statSync(f).isFile()) return d;
      const parent = dirname(d);
      if (parent === d) break;
      d = parent;
    }
  }
  return null;
}

function read_text(path) {
  return readFileSync(path, "utf-8");
}

function footer_line_index(lines, fid) {
  const esc = fid.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const pat = new RegExp(`<!--\\s*id:\\s*${esc}\\s*\\|`);
  for (let i = 0; i < lines.length; i++) if (pat.test(lines[i])) return i;
  return null;
}

function block_top(lines, footer_idx) {
  // walk back from the footer to the nearest column-0 list bullet ('- ')
  for (let i = footer_idx; i >= 0; i--) if (lines[i].startsWith("- ")) return i;
  return footer_idx;
}

function derive_note(block_lines) {
  let first = block_lines[0];
  first = first.replace(/^- \[[ xX]\]\s*/, "");
  first = first.replace(/^- \s*/, "");
  first = first.replaceAll("**", "").trim();
  first = first.replace(/\s+/g, " ");
  return first.length > 90 ? first.slice(0, 90).replace(/\s+$/, "") + "…" : first;
}

function derive_quarter(y, mo) {
  return `${y}-Q${Math.floor((mo - 1) / 3) + 1}`;
}

export function archive_facts(root, ids, reason, quarter, note, dry_run) {
  const mem = join(root, "memory");
  const cont_path = join(mem, "continuity.md");
  const lines = read_text(cont_path).split("\n");

  const archived_ids = new Set();
  const arch_dir = join(mem, "archive");
  if (existsSync(arch_dir)) {
    for (const name of readdirSync(arch_dir)) {
      if (name.endsWith(".md") && !name.toUpperCase().startsWith("INDEX")) {
        for (const m of read_text(join(arch_dir, name)).matchAll(/<!--\s*id:\s*([a-z0-9-]+)\s*\|/g)) {
          archived_ids.add(m[1]);
        }
      }
    }
  }

  const plans = [];
  for (const fid of ids) {
    if (archived_ids.has(fid)) return { code: 1, msg: `refused: '${fid}' is already in the archive — nothing moved` };
    const fidx = footer_line_index(lines, fid);
    if (fidx === null) return { code: 1, msg: `refused: '${fid}' has no footer in continuity.md — nothing moved` };
    plans.push([fid, block_top(lines, fidx), fidx]);
  }

  const plans_sorted = [...plans].sort((a, b) => b[1] - a[1]); // bottom-up removal

  const now = new Date();
  quarter = quarter || derive_quarter(now.getUTCFullYear(), now.getUTCMonth() + 1);
  const isoDate = now.toISOString().slice(0, 10);
  const arch_path = join(mem, "archive", `${quarter}.md`);
  const index_path = join(mem, "archive", "INDEX.md");

  const moved = [];
  const archive_chunks = [];
  const index_chunks = [];
  const kept = [...lines];
  for (const [fid, top, bot] of plans_sorted) {
    const block = lines.slice(top, bot + 1);
    const n = note && ids.length === 1 ? note : derive_note(block);
    moved.push([fid, n]);
    archive_chunks.push(block.join("\n"));
    index_chunks.push(`- ${fid} — ${n} — ${reason} — ${quarter}.md`);
    let end = bot + 1;
    if (end < kept.length && kept[end].trim() === "") end += 1;
    kept.splice(top, end - top);
  }
  moved.reverse();
  archive_chunks.reverse();
  index_chunks.reverse();

  const heading = `\n## ${isoDate} — archived via archive-fact (${reason})\n\n### Faded facts\n\n`;
  const archive_block = heading + archive_chunks.join("\n\n") + "\n";
  const index_block = index_chunks.join("\n") + "\n";
  const new_cont = kept.join("\n");

  if (new_cont.trim() === "") return { code: 1, msg: "refused: removing those blocks would empty continuity.md — nothing moved" };

  const summary = moved.map(([fid, n]) => `  ${fid} → ${quarter}.md  (${n})`).join("\n");
  if (dry_run) return { code: 0, msg: `DRY-RUN — would move ${moved.length} fact(s):\n${summary}\n(no files changed)` };

  appendFileSync(arch_path, archive_block, "utf-8");
  appendFileSync(index_path, index_block, "utf-8");
  writeFileSync(cont_path, new_cont, "utf-8"); // safe: new_cont already in memory

  return { code: 0, msg: `moved ${moved.length} fact(s) to ${quarter}.md + INDEX:\n${summary}\nNow run memory-lint to confirm.` };
}

function parse_args(argv) {
  const out = { ids: [], reason: "faded", quarter: null, note: null, root: null, dry_run: false };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--id") out.ids.push(argv[++i]);
    else if (a === "--reason") out.reason = argv[++i];
    else if (a === "--quarter") out.quarter = argv[++i];
    else if (a === "--note") out.note = argv[++i];
    else if (a === "--root") out.root = argv[++i];
    else if (a === "--dry-run") out.dry_run = true;
  }
  return out;
}

export function main(argv) {
  const args = parse_args(argv ?? process.argv.slice(2));
  if (args.ids.length === 0) {
    console.error("archive-fact: --id is required");
    return 2;
  }
  const root = find_root(args.root || process.cwd());
  if (!root) {
    console.error("archive-fact: could not find memory/continuity.md");
    return 2;
  }
  const { code, msg } = archive_facts(root, args.ids, args.reason, args.quarter, args.note, args.dry_run);
  const line = "archive-fact: " + msg;
  if (code === 0) console.log(line);
  else console.error(line);
  return code;
}

if (process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  process.exit(main());
}

File: agent-skills/archive-fact/scripts/archive-fact.py
#!/usr/bin/env python3
"""archive-fact — safely MOVE a decayed fact from continuity.md to the archive.

This executes the mechanical half of REVIEW.md step 4 deterministically, so the
move can't be botched. The agent still decides *what* to archive (meaning); this
script performs the *move* safely (mechanics) — the same split as memory-lint.

Why it exists: the read-modify-write `open(f,"w").write(open(f).read()+…)` truncates
the file to empty BEFORE the read runs. That exact antipattern wiped a version.md
stamp and then this repo's archive (50 facts → 6), twice. This helper reads the
whole file into memory FIRST and writes once, so truncation is structurally impossible.

Usage:
    python3 archive-fact.py --id ID [--id ID2 ...] [--reason REASON]
                            [--quarter YYYY-Qn] [--note TEXT] [--root PATH] [--dry-run]

  --reason   "faded" (default) or e.g. "superseded by new-id"
  --quarter  archive file memory/archive/<quarter>.md (default: derived from today UTC)
  --note     INDEX one-liner override (default: derived from the block's first line)
  --dry-run  print what would move; change nothing

Exit: 0 = moved (or dry-run ok), 1 = refused (id missing / already archived / unsafe),
2 = could not locate the memory/ layer. Run `memory-lint` afterward to confirm.
"""
import argparse
import datetime
import os
import re
import sys


def find_root(start):
    for cand in (start, os.getcwd(), os.path.dirname(os.path.abspath(__file__))):
        if not cand:
            continue
        d = os.path.abspath(cand)
        while True:
            if os.path.isfile(os.path.join(d, "memory", "continuity.md")):
                return d
            parent = os.path.dirname(d)
            if parent == d:
                break
            d = parent
    return None


def read_text(path):
    with open(path, encoding="utf-8") as f:
        return f.read()


def footer_line_index(lines, fid):
    """Index of the single-line footer for fid, or None."""
    pat = re.compile(r"<!--\s*id:\s*" + re.escape(fid) + r"\s*\|")
    for i, ln in enumerate(lines):
        if pat.search(ln):
            return i
    return None


def block_span(lines, footer_idx):
    """[top, footer] inclusive: walk back from the footer to the nearest column-0
    list bullet ('- '), which starts the fact/thread block."""
    top = footer_idx
    for i in range(footer_idx, -1, -1):
        if lines[i].startswith("- "):
            top = i
            break
    return top, footer_idx


def derive_note(block_lines):
    first = block_lines[0]
    first = re.sub(r"^- \[[ xX]\]\s*", "", first)  # drop checkbox
    first = re.sub(r"^- \s*", "", first)
    first = first.replace("**", "").strip()
    first = re.sub(r"\s+", " ", first)
    return (first[:90].rstrip() + "…") if len(first) > 90 else first


def derive_quarter(today):
    return f"{today.year}-Q{(today.month - 1) // 3 + 1}"


def archive_facts(root, ids, reason, quarter, note, dry_run):
    mem = os.path.join(root, "memory")
    cont_path = os.path.join(mem, "continuity.md")
    cont = read_text(cont_path)
    lines = cont.split("\n")

    # Archived-already guard: scan existing archive footers (any *.md but INDEX).
    archived_ids = set()
    arch_dir = os.path.join(mem, "archive")
    if os.path.isdir(arch_dir):
        for name in os.listdir(arch_dir):
            if name.endswith(".md") and not name.upper().startswith("INDEX"):
                for m in re.finditer(r"<!--\s*id:\s*([a-z0-9-]+)\s*\|", read_text(os.path.join(arch_dir, name))):
                    archived_ids.add(m.group(1))

    # Validate ALL ids before touching anything (all-or-nothing).
    plans = []
    for fid in ids:
        if fid in archived_ids:
            return 1, f"refused: '{fid}' is already in the archive — nothing moved"
        fidx = footer_line_index(lines, fid)
        if fidx is None:
            return 1, f"refused: '{fid}' has no footer in continuity.md — nothing moved"
        top, bot = block_span(lines, fidx)
        plans.append((fid, top, bot))

    # Order removals bottom-up so earlier indices stay valid.
    plans_sorted = sorted(plans, key=lambda p: p[1], reverse=True)

    today = datetime.datetime.now(datetime.timezone.utc).date()
    quarter = quarter or derive_quarter(today)
    arch_path = os.path.join(mem, "archive", f"{quarter}.md")
    index_path = os.path.join(mem, "archive", "INDEX.md")

    moved = []
    archive_chunks = []
    index_chunks = []
    kept = list(lines)
    for fid, top, bot in plans_sorted:
        block = lines[top : bot + 1]
        n = note if (note and len(ids) == 1) else derive_note(block)
        moved.append((fid, n))
        archive_chunks.append("\n".join(block))
        index_chunks.append(f"- {fid} — {n} — {reason} — {quarter}.md")
        # remove block + one trailing blank line if present
        end = bot + 1
        if end < len(kept) and kept[end].strip() == "":
            end += 1
        del kept[top:end]

    moved.reverse()
    archive_chunks.reverse()
    index_chunks.reverse()

    heading = (
        f"\n## {today.isoformat()} — archived via archive-fact ({reason})\n\n### Faded facts\n\n"
    )
    archive_block = heading + "\n\n".join(archive_chunks) + "\n"
    index_block = "\n".join(index_chunks) + "\n"
    new_cont = "\n".join(kept)

    if not new_cont.strip():
        return 1, "refused: removing those blocks would empty continuity.md — nothing moved"

    summary = "\n".join(f"  {fid} → {quarter}.md  ({n})" for fid, n in moved)
    if dry_run:
        return 0, f"DRY-RUN — would move {len(moved)} fact(s):\n{summary}\n(no files changed)"

    # WRITES — append-mode for the archive/INDEX (never truncates); single write for continuity.
    with open(arch_path, "a", encoding="utf-8") as f:
        f.write(archive_block)
    with open(index_path, "a", encoding="utf-8") as f:
        f.write(index_block)
    with open(cont_path, "w", encoding="utf-8") as f:  # safe: new_cont already in memory
        f.write(new_cont)

    return 0, f"moved {len(moved)} fact(s) to {quarter}.md + INDEX:\n{summary}\nNow run memory-lint to confirm."


def main(argv=None):
    ap = argparse.ArgumentParser(description="Safely move a decayed fact from continuity.md to the archive.")
    ap.add_argument("--id", action="append", dest="ids", required=True, help="fact id (repeatable)")
    ap.add_argument("--reason", default="faded", help='"faded" (default) or "superseded by <id>"')
    ap.add_argument("--quarter", default=None, help="archive quarter file, e.g. 2026-Q2 (default: today)")
    ap.add_argument("--note", default=None, help="INDEX one-liner (single --id only; else derived)")
    ap.add_argument("--root", default=None)
    ap.add_argument("--dry-run", action="store_true")
    args = ap.parse_args(argv)

    root = find_root(args.root or os.getcwd())
    if not root:
        print("archive-fact: could not find memory/continuity.md", file=sys.stderr)
        return 2
    code, msg = archive_facts(root, args.ids, args.reason, args.quarter, args.note, args.dry_run)
    print(("archive-fact: " + msg) if code == 0 else ("archive-fact: " + msg), file=(sys.stdout if code == 0 else sys.stderr))
    return code


if __name__ == "__main__":
    sys.exit(main())
